MAINT: linalg: Empty array argument support follow-up

JozsefKutas opened the pull request:
#### Reference issue
Contributes to #17658, follow-up on #18141

#### What does this implement/fix?
This adds support for empty matrices for several `linalg` functions: `inv`, `pinv`. `pinvh`, `eig`, `eigh`, `svd`, `schur`, building on work done by @boatwrong. I've ammended the PR so that the return values have the correct dimensions and take into account optional function arguments, as described in my comment on #18141.

If people think this approach makes sense, I can work on expanding support to other linear algebra functions.

JozsefKutas commented:
Re. `scipy.linalg.kron`, I'm not sure whether it might not be better to just delegate to `numpy.kron`, unless there are any advantages to `scipy.linalg.kron`. `numpy.kron` works with empty array arguments and also handles n-dimensional arrays, whereas `scipy.linalg.kron` gives some funny results when the number of dimensions is not two:

```python
linalg.kron(np.empty((1,2,3)), np.empty((4,5,6))).shape
(3, 8, 5, 6)  # should be (4, 10, 18)
```

ilayn commented:
By the way, det and lu are already handling empty inputs. Above your tests there are bunch of empty array tests

JozsefKutas commented:
Thanks for spotting that, I'll remove the `det` test. I didn't modify the tests for LU - I saw you had worked on that already. I added tests for `lu_factor` and `lu_solve`.

JozsefKutas commented:
This should cover all basic linear algebra functions and decompositions apart from `qz` and `ordqz` (these might require some refactoring). I noticed a few documentation issues while working on this, but I'll put them in a separate PR.

JozsefKutas commented:
Some explanations and thoughts:

- For the more complicated decomposition functions I added parametric tests checking that each function returns arrays with the correct types and shapes when different options are used, as well as the empty array input tests. These check that the results for empty arrays are consistent with the results for non-empty arrays.

- While updating test cases, I removed `test_misc.py`, as there didn't seem to be anything there.

- I check for empty arrays using the condition `a1.size == 0`, but I noticed in `det` we use `min(*a1.shape) == 0`. Are there any advantages to this approach? As far as I can see this only makes a difference for 0-d arrays, which should usually be caught in an initial shape check.

- Argument validations vary from function to function, and in some cases I'm not 100% sure which values should be allowed and tested against. For example, `solve` accepts n-d arrays, though this isn't mentioned in the documentation and the results are not consistent with `np.linalg.solve`.

- In general LAPACK should handle cases where dimensions passed as arguments are zero, but it requires most array inputs to have size at least one. This generally can't be achieved through scipy's LAPACK functions.

This ended up being a fair bit more work than I anticipated when I started, and there's a fair amount of new code to handle and test different cases. I can split this into multiple PRs if that makes the review process easier.

ev-br commented:
When this lands, eeb1bac3ce3da33e6436cb01ce1d974ad4badf00 should be attributed back to @boatwrong .

ev-br commented on scipy/linalg/tests/test_basic.py:
```diff
@@ -174,6 +174,17 @@ def test_native_list_arguments(self):
+        ab = [[]]
+        b = []
+        x = solve_banded((0, 0), ab, b)
+        assert_allclose(x, [])
```
```
In [1]: import numpy as np

In [2]: a = np.empty((0, 0))

In [3]: from numpy.testing import assert_allclose

In [4]: assert_allclose(a, 8)     # does not fail!
```

What does the assert_allclose actually check here.
ISTM the check should be `x.shape == (0,)`, what am I missing?

JozsefKutas replied:
I don't think you're missing anything. I guess it didn't occur to me that any of these functions might return a scalar and that this would bypass the shape check in `assert_allclose`. A direct shape check would make more sense. It's been a while since I last looked at this, but I can amend this and similar assertions.

ev-br replied:
Yeah, I fully realize it must be rather frustrating to be asked to do a major change this late! 
Would be very welcome though. My intention is to review/land this PR "soon", and this comment is me starting to work out what is the endgame.

ev-br replied:
Actually, it's not just shape, it's also dtype:

```
In [5]: a32 = np.eye(0, dtype=np.float32)

In [6]: np.linalg.cholesky(a32)
Out[6]: array([], shape=(0, 0), dtype=float32)
```

I _suspect_ the majority of the code itself it fine because of `np.empty_like`, but this definitely needs testing.

j-bowhay commented:
Closing in favor of #20295
